### PR TITLE
Add `StreamMessage.decode()`

### LINKED ISSUE #4147: Add `StreamMessage.decode()`
It'd be nice if a user can make use of `HttpDecoder` for `StreamMessage` as well as `Http{Request,Response}`. We could:

- Add `StreamDecoder`
- Make `HttpDecoder` extend `StreamDecoder` with a few additional methods such as `processHeaders()`
- Rename `HttpDecoder{Input,Output}` to `StreamDecoder{Input,Output}`

Additionally, `StreamDecoder.processOnComplete()` could have `StreamDecoderInput` so it can handle the remaining data.

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

`HttpDecoder` can use with `HttpMessage` not `StreamMessage`.
It would be nice to add support for decoding `StreamMessage` itself.
See #4147 for details.

Modifications:

- Add `StreamDecoder` and move up `process()`, `processOnComplete()` and
  `processOnError()` from `HttpDecoder`.
  - `decodeInput()` is added to `StreamDecoder` for converting an
    arbitrary object to `ByteBuf`.
- Change `ArmeriaMessageDeframer` to take `grpcWebText` flag to
  decode a Base64-encoded `HttpData`.
- Fix various gRPC implementations to leverage new APIs.
- Breaking)
  - Rename `HttpDecoder{Input,Output}` to `StreamDecoder{Input,Output}`.
  - `HttpMessage.decode(..., Function<? super HttpData, ? extends ByteBuf>)`
    has been removed in favor of `decodedInput()`.
  - `HttpDecoder.proccessOnComplete()` now takes `StreamDecoderInput` as
    the first parameter.

Result:

- You can now decode a `StreamMessage` using `StreamMessage.decode()`.
- Fixes #4147

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamDecoder.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+ * @param <O> the output type of being decoded
+ */
+@UnstableApi
+public interface StreamDecoder<I, O> {
+
+    /**
+     * Converts the specified {@link I} type object into a {@link ByteBuf} that is added to
+     * {@link StreamDecoderInput}.
+     */
+    ByteBuf decodeInput(I in);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamDecoder.java:92]: How about `toByteBuf()`? Its purpose doesn't look like decoding something. Any suggestions for a better name would be appreciated.

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/DecodedStreamMessage.java [hunk 2]
```diff
@@ -185,7 +189,7 @@ private void cleanup() {
         input.close();
     }
 
-    private final class HttpMessageSubscriber implements Subscriber<HttpObject> {
+    private final class MessageSubscriber implements Subscriber<I> {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/common/stream/DecodedStreamMessage.java:192]: `DecodingSubscriber`?

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamDecoder.java (hunk 1 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/stream/StreamDecoder.java:92]: > HttpData.byteBuf() and then do the actual decoding in process()

Makes sense. Let me use `toByteBuf()`. To be honest, that was my initial thought.

**Review by trustin — APPROVED** [review]
:+1:

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamDecoder.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    ByteBuf toByteBuf(I in);
+
+    /**
+     * Decodes a stream of data to N objects.
+     * This method will be called whenever an object is signaled from {@link Publisher}.
+     */
+    void process(StreamDecoderInput in, StreamDecoderOutput<O> out) throws Exception;
+
+    /**
+     * Invoked when {@link HttpData}s are fully consumed.
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamDecoder.java:101]:
```suggestion
     * Invoked when the input is fully consumed.
```

**Review by jrhee17 — APPROVED** [review]
Left a super nit comment. Looks like the best way to support this feature! Thanks @ikhoon 👍 🙇 👍

**Pull request merged**